[dashboard agent] Catch agent port conflict

SongGuyang opened the pull request:
## Why are these changes needed?
- Catch all the port conflicts in agent and stay agent alive. There are three port:
    * grpc port
    * metric exporter port
    * http listening port
- Turn on `raylet_shares_fate_with_agent` by default.
- Remove the agent restart feature.
- Refer to issue https://github.com/ray-project/ray/issues/21695#issuecomment-1032161528


## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

rkooo567 commented:
Minimal install failures seesm related

rkooo567 commented on dashboard/agent.py:
```diff
@@ -159,7 +173,13 @@ async def _check_parent():
+            try:
+                self.http_server = await self._configure_http_server(modules)
+            except Exception:
+                logger.exception(
```
Can you add TODO here to remove this in the future? (and do better port resolution)

rkooo567 commented on dashboard/agent.py:
```diff
@@ -361,36 +381,27 @@ async def _check_parent():
-                gcs_publisher=gcs_publisher,
-            )
-            logger.error(message)
+        # Agent is failed to be started many times.
```
This comment seems wrong?

rkooo567 commented on dashboard/agent.py:
```diff
@@ -361,36 +381,27 @@ async def _check_parent():
+        gcs_publisher = GcsPublisher(address=args.gcs_address)
+
+        traceback_str = ray._private.utils.format_error_message(traceback.format_exc())
+        message = (
```
Do we need it? In this case, raylet will just fail right?

SongGuyang replied:
Yeah, if we open fate sharing by default, we don't need this. I can remove this.

rkooo567 commented on dashboard/modules/reporter/reporter_agent.py:
```diff
@@ -603,7 +603,8 @@ async def _perform_iteration(self, publisher):
 
     async def run(self, server):
-        reporter_pb2_grpc.add_ReporterServiceServicer_to_server(self, server)
+        if server:
```
Any way to just not run this function instead of adding this logic? Seems a bit ugly

SongGuyang replied:
I didn't find better way.  The `run` is a common interface of dashboard agent module. We can do anything else in this function, besides starting grpc server. So, if we don't run this function, we could lose some code paths or features.

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -121,8 +121,6 @@ def test_port_conflict(call_ray_stop_only, shutdown_only):
     with pytest.raises(ValueError, match="already occupied"):
         ray.init(dashboard_port=9999, include_dashboard=True)
 
-    sock.close()
```
Why removing this?

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -161,48 +159,107 @@ def set_agent_failure_env_var():
-    ],
-    indirect=True,
-)
 def test_dashboard_agent_restart(
```
Can you change the name of thet test?

rkooo567 replied:
Let's just remove it and instead add a test to verify fate sharing

SongGuyang commented:
Yeah, not sure the difference of minimal env. I should improve or modify the test.

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -161,48 +159,107 @@ def set_agent_failure_env_var():
     # Make sure there's no spammy message for 5 seconds.
     def matcher(log_batch):
-        return log_batch["pid"] != "autoscaler"
+        return log_batch["pid"] == "raylet" and any(
```
Can you also check if the raylet is also killed?

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -161,48 +159,107 @@ def set_agent_failure_env_var():
+conflict_port = 34567
+
+
+def run_tasks_without_runtime_env():
```
Can you add assert ray.is_initialized()

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -161,48 +159,107 @@ def set_agent_failure_env_var():
+
+
+def run_tasks_with_runtime_env():
+    @ray.remote(runtime_env={"pip": ["pip-install-test==0.5"]})
```
Same as above; 

Can you add assert ray.is_initialized()

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -161,48 +159,107 @@ def set_agent_failure_env_var():
+    # Tasks without runtime env still work when dashboard agent grpc port conflicts.
+    run_tasks_without_runtime_env()
+    # Tasks with runtime env couldn't work.
+    with pytest.raises(ray.exceptions.RuntimeEnvSetupError):
```
Can you make sure the error message contains the issue is from the port conflict?

SongGuyang replied:
Good! done

rkooo567 commented on python/ray/tests/test_runtime_env.py:
```diff
@@ -330,13 +318,13 @@ def f():
     Test task raises an exception.
     """
-    with pytest.raises(RuntimeEnvSetupError):
+    with pytest.raises(ray.exceptions.LocalRayletDiedError):
```
Can you check there's the error message that says the reason is due to agent failure?

SongGuyang replied:
I think I can remove this test. In this test(`test_runtime_env_broken`), we injected a env to make agent failed. Agent failure will lead to raylet failure. Raylet failure will lead to driver failure(raylet client failed). We can't catch any error message.

rkooo567 commented on python/ray/tests/test_runtime_env.py:
```diff
@@ -330,13 +318,13 @@ def f():
     Test actor task raises an exception.
     """
     a = A.options(runtime_env=runtime_env).remote()
-    with pytest.raises(ray.exceptions.RuntimeEnvSetupError):
```
Same as above

rkooo567 commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -32,13 +32,19 @@ void AgentManager::HandleRegisterAgent(const rpc::RegisterAgentRequest &request,
+    RAY_LOG(INFO) << "HandleRegisterAgent, ip: " << agent_ip_address_
+                  << ", port: " << agent_port_ << ", pid: " << agent_pid_;
+  } else {
+    RAY_LOG(ERROR) << "The grpc port of agent is invalid (be 0), ip: "
```
Can you use WARNING instead? ERROR will propagate to the driver

rkooo567 commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -162,10 +142,11 @@ void AgentManager::CreateRuntimeEnv(
   if (runtime_env_agent_client_ == nullptr) {
     // If the agent cannot be restarted anymore, fail the request.
-    if (agent_restart_count_ >= RayConfig::instance().agent_max_restart_count()) {
+    if (disable_agent_client_ || agent_failed_) {
```
If agent_failed == true, this line should never reach? Maybe instead

```
RAY_CHECK(!agent_failed)
if (disable_agent_client_)
```

SongGuyang replied:
Rmove the flag `agent_failed_`.

rkooo567 commented on src/ray/raylet/agent_manager.h:
```diff
@@ -87,15 +87,16 @@ class AgentManager : public rpc::AgentManagerServiceHandler {
   DelayExecutorFn delay_executor_;
   RuntimeEnvAgentClientFactoryFn runtime_env_agent_client_factory_;
   std::shared_ptr<rpc::RuntimeEnvAgentClientInterface> runtime_env_agent_client_;
+  bool disable_agent_client_ = false;
```
Can you add a comment to explain the variable?

rkooo567 commented:
Most of comments are about making sure we have the right message when things failed due to this issue

SongGuyang commented:
Thanks! I will address the comments(including another PR) tonight and ping you guys again.

SongGuyang commented:
@rkooo567 Please take a look agent. One question is that can I skip the two tests? https://github.com/ray-project/ray/pull/23024/commits/32ce0e2dded943949e8c2e7b8d80e56c88d6ddad

rkooo567 commented on python/ray/tests/test_dashboard.py:
```diff
@@ -161,48 +159,118 @@ def set_agent_failure_env_var():
+    # except grpc port.
+    run_tasks_with_runtime_env()
+
+
```
I think we don't have a test to verify fate sharing?

rkooo567 commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -32,13 +32,19 @@ void AgentManager::HandleRegisterAgent(const rpc::RegisterAgentRequest &request,
-      runtime_env_agent_client_factory_(agent_ip_address_, agent_port_);
-  RAY_LOG(INFO) << "HandleRegisterAgent, ip: " << agent_ip_address_
-                << ", port: " << agent_port_ << ", pid: " << agent_pid_;
+  // Note: `agent_port_` should be 0 if the grpc port of agent is in conflict.
```
Can you add TODO here to remove this workaround?

rkooo567 reviewed (commented):
https://github.com/ray-project/ray/blob/2294a7ed4745cd11a0da575cd1902035ec6d7afd/src/ray/raylet/agent_manager.cc#L223

-> We should also handle this better. Maybe just fail the request instead of running the infinite retry? It should be fine since the agent will fate share with raylet

rkooo567 commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -230,6 +208,11 @@ void AgentManager::CreateRuntimeEnv(
 void AgentManager::DeleteURIs(const std::vector<std::string> &uris,
                               DeleteURIsCallback callback) {
+  if (disable_agent_client_) {
+    RAY_LOG(ERROR) << "Failed to delete URIs because the agent client is disabled.";
```
Since it is the user facing error, we should refine it a bit more.

Based on https://spark.apache.org/error-message-guidelines.html

```
Failed to delete runtime environment URI because the ray agent couldn't be started due to the port conflict. See `dashboard_agent.log` for more details. To solve the problem, start ray with a hard-coded agent port. `ray start --dashboard-agent-grpc-port [port]` and make sure the port is not used by other processes.
```

- Who encountered the problem?
What was the problem?
When did the problem happen?
Where did the problem happen?
Why did the problem happen?
How can the problem be solved?

SongGuyang commented:
We already have the fate sharing test named `test_raylet_and_agent_share_fate`.

About the driver failure, the error message is:
```
::test_runtime_env_broken[dict] 2022-03-11 08:33:06,463	INFO worker.py:978 -- Connecting to existing Ray cluster at address: 172.18.0.3:59955
--
  | [2022-03-11 08:33:15,674 E 6912 13117] core_worker_process.cc:218: Failed to get the system config from raylet because it is dead. Worker will terminate. Status: GrpcUnavailable: RPC Error message: failed to connect to all addresses; RPC Error details:
```
I saw this in the https://buildkite.com/ray-project/ray-builders-pr/builds/26582#4eecbf4d-a236-4ab9-b350-10dfd971fb04

rkooo567 commented:
> We already have the fate sharing test named `test_raylet_and_agent_share_fate`.
> 
> About the driver failure, the error message is:
> 
> ```
> ::test_runtime_env_broken[dict] 2022-03-11 08:33:06,463	INFO worker.py:978 -- Connecting to existing Ray cluster at address: 172.18.0.3:59955
> --
>   | [2022-03-11 08:33:15,674 E 6912 13117] core_worker_process.cc:218: Failed to get the system config from raylet because it is dead. Worker will terminate. Status: GrpcUnavailable: RPC Error message: failed to connect to all addresses; RPC Error details:
> ```
> 
> I saw this in the https://buildkite.com/ray-project/ray-builders-pr/builds/26582#4eecbf4d-a236-4ab9-b350-10dfd971fb04

Hmm. this one seems to be a pretty bad error message... Can you at least mention to the error message
```
core_worker_process.cc:218: Failed to get the system config from raylet because it is dead. Worker will terminate.
```
to take a look at raylet.out log file? I think we should find a better way to propagate error messages, but it seems to be difficult to achieve it in the short term

rkooo567 reviewed (approved):
LGTM given the remaining comment are addressed.

Btw, do you think this should be a part of 1.12? My impression is it could be a breaking change, and having enough time (1 release) seems to be not a bad choice. Do you guys have hard dependencies on this feature?

SongGuyang commented:
> Also do you happen to know why this only happens in minimal install, not a regular test?

I don't know the reason. Maybe the test is finished before driver detects the raylet failure.

SongGuyang commented:
> Btw, do you think this should be a part of 1.12? My impression is it could be a breaking change, and having enough time (1 release) seems to be not a bad choice. Do you guys have hard dependencies on this feature?

I think sharing fate, catching port conflict, and URI reference refactor can enhance the stability, instead of regression. Do we also have 1 or 2 weeks to test before 1.12 release? We have two choices:

option 1: Merge current PR and https://github.com/ray-project/ray/pull/22828 before the cut of 1.12. And we should do more test for this before 1.12 release.

option 2: Merge current PR and https://github.com/ray-project/ray/pull/22828 after the cut of 1.12. But we should merge it as soon as possible because it blocks a lot of following PRs.

I tend to option 1, but it's also ok to choose option 2. @edoakes @architkulkarni What do you think?

kfstorm commented:
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

architkulkarni commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -32,13 +32,20 @@ void AgentManager::HandleRegisterAgent(const rpc::RegisterAgentRequest &request,
+    RAY_LOG(INFO) << "HandleRegisterAgent, ip: " << agent_ip_address_
+                  << ", port: " << agent_port_ << ", pid: " << agent_pid_;
+  } else {
+    RAY_LOG(WARNING) << "The grpc port of agent is invalid (be 0), ip: "
+                     << agent_ip_address_ << ", pid: " << agent_pid_
+                     << ". Disable the agent client in raylet.";
```
```suggestion
    RAY_LOG(WARNING) << "The gRPC port of the Ray agent is invalid (0), ip: "
                     << agent_ip_address_ << ", pid: " << agent_pid_
                     << ". The agent client in the raylet has been disabled.";
```

architkulkarni commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -102,35 +104,11 @@ void AgentManager::StartAgent() {
-        QuickExit();
-      }
-    }
+    RAY_LOG(ERROR) << "Raylet exits immediately because the ray agent has failed. "
+                      "Raylet fate shares with the agent. It can happen because the "
+                      "Ray agent is unexpectedly killed or failed. See "
+                      "`dashboard_agent.log` for the root cause.";
```
```suggestion
    RAY_LOG(ERROR) << "The raylet exited immediately because the Ray agent failed. "
                      "The raylet fate shares with the agent. This can happen because the "
                      "Ray agent was unexpectedly killed or failed. See "
                      "`dashboard_agent.log` for the root cause.";
```

architkulkarni commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -163,19 +141,20 @@ void AgentManager::CreateRuntimeEnv(
-      str_stream << "Runtime environment " << serialized_runtime_env
-                 << " cannot be created on this node because the agent is dead.";
+      str_stream
+          << "Runtime environment " << serialized_runtime_env
+          << " cannot be created on this node because the agent client is disabled. You "
+             "see this error message maybe because the grpc port of agent came into "
+             "conflict. Please see `dashboard_agent.log` to get more details.";
```
```suggestion
          << "The runtime environment " << serialized_runtime_env
          << " cannot be created on this node because the agent client is disabled. You "
             "might see this error message because the gRPC port of the agent came into "
             "conflict. Please see `dashboard_agent.log` to get more details.";
```

architkulkarni commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -226,29 +203,25 @@ void AgentManager::CreateRuntimeEnv(
                               DeleteURIsCallback callback) {
+  if (disable_agent_client_) {
+    RAY_LOG(ERROR)
+        << "Failed to delete runtime env URIs because the agent client is disabled. You "
+           "see this error message maybe because the grpc port of agent came into "
+           "conflict. Please see `dashboard_agent.log` to get more details.";
```
```suggestion
        << "Failed to delete runtime env URIs because the agent client is disabled. You "
           "might see this error message because the gRPC port of the agent came into "
           "conflict. Please see `dashboard_agent.log` to get more details.";
```

architkulkarni reviewed (approved):
Looks good to me, just added some nits about the wording in some of the error messages (I know some of them predated this PR, it would be nice to add them but no need to block the PR on this)

One more comment though, can we add in Sang's suggestion to the error messages? 

> To solve the problem, start ray with a hard-coded agent port. `ray start --dashboard-agent-grpc-port [port]` and make sure the port is not used by other processes.
`. 

I think it would be really helpful for users who see the error message.

architkulkarni commented:
> > Btw, do you think this should be a part of 1.12? My impression is it could be a breaking change, and having enough time (1 release) seems to be not a bad choice. Do you guys have hard dependencies on this feature?
> 
> I think sharing fate, catching port conflict, and URI reference refactor can enhance the stability, instead of regression. Do we also have 1 or 2 weeks to test before 1.12 release? We have two choices:
> 
> option 1: Merge current PR and #22828 before the cut of 1.12. And we should do more test for this before 1.12 release.
> 
> option 2: Merge current PR and #22828 after the cut of 1.12. But we should merge it as soon as possible because it blocks a lot of following PRs.
> 
> I tend to option 1, but it's also ok to choose option 2. @edoakes @architkulkarni What do you think?

I agree that the stability will be improved by these two PRs, so I think option 1 would be better.

rkooo567 commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -206,9 +185,7 @@ void AgentManager::CreateRuntimeEnv(
-      [this,
-       job_id,
-       serialized_runtime_env,
+      [serialized_runtime_env,
```
```suggestion
      [serialized_runtime_env=std::move(serialized_runtime_env),
```
Is it possible to do this?

SongGuyang replied:
We can't do this because `serialized_runtime_env` is a reference from task_spec.

rkooo567 commented on src/ray/raylet/agent_manager.cc:
```diff
@@ -261,7 +234,7 @@ void AgentManager::DeleteURIs(const std::vector<std::string> &uris,
   }
   runtime_env_agent_client_->DeleteURIs(
-      request, [this, uris, callback](Status status, const rpc::DeleteURIsReply &reply) {
+      request, [callback](Status status, const rpc::DeleteURIsReply &reply) {
```
```suggestion
      request, [callback=std::move(callback)](Status status, const rpc::DeleteURIsReply &reply) {
```
?

SongGuyang replied:
It's ok

SongGuyang commented:
> Looks good to me, just added some nits about the wording in some of the error messages (I know some of them predated this PR, it would be nice to add them but no need to block the PR on this)
> 
> One more comment though, can we add in Sang's suggestion to the error messages?
> 
> > To solve the problem, start ray with a hard-coded agent port. `ray start --dashboard-agent-grpc-port [port]` and make sure the port is not used by other processes.
> > `.
> 
> I think it would be really helpful for users who see the error message.

Good comments! Fixed!
